File: everything_parsing/29-llk-adaptive/java/Backtrack.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Module 29 — a backtracking recursive-descent recognizer that COUNTS.
 *
 * `deriveSym` tries every production of a nonterminal in order, following all of
 * them (no lookahead pruning). It returns, for each end position, the number of
 * distinct ways to derive the consumed span — so the total for
 * position == input.length is the number of parse trees.
 *
 * `entries` counts production attempts: the raw work. With no memoization this
 * is exponential for a grammar whose alternatives overlap (see equiv.grammar).
 * Packrat parsing (Appendix X2) memoizes (nonterminal, position) to make it
 * linear.
 *
 * Grammars must be free of left recursion (checked by the caller) or this
 * recurses forever.
 */
public final class Backtrack {

    public static final class Capped extends RuntimeException {}

    public record Result(long parseCount, long entries, boolean capped) {}

    private final LeftRec.G g;
    private final long cap;
    private long entries;

    public Backtrack(LeftRec.G g) { this(g, 20_000_000L); }
    public Backtrack(LeftRec.G g, long cap) { this.g = g; this.cap = cap; }

    public Result parse(List<String> input) {
        entries = 0;
        try {
            Map<Integer, Long> ends = deriveSym(g.start, 0, input);
            long full = ends.getOrDefault(input.size(), 0L);
            return new Result(full, entries, false);
        } catch (Capped c) {
            return new Result(-1, entries, true);
        }
    }

    private Map<Integer, Long> deriveSym(String sym, int pos, List<String> input) {
        if (!g.isNT(sym)) {
            Map<Integer, Long> out = new HashMap<>();
            if (pos < input.size() && input.get(pos).equals(sym)) out.put(pos + 1, 1L);
            return out;
        }
        Map<Integer, Long> out = new HashMap<>();
        for (List<String> rhs : g.prods.get(sym)) {
            if (++entries > cap) throw new Capped();
            Map<Integer, Long> r = deriveSeq(rhs, pos, input);
            for (var e : r.entrySet()) out.merge(e.getKey(), e.getValue(), Long::sum);
        }
        return out;
    }

    private Map<Integer, Long> deriveSeq(List<String> seq, int pos, List<String> input) {
        Map<Integer, Long> cur = new HashMap<>();
        cur.put(pos, 1L);
        for (String sym : seq) {
            Map<Integer, Long> nxt = new HashMap<>();
            for (var pe : cur.entrySet()) {
                Map<Integer, Long> r = deriveSym(sym, pe.getKey(), input);
                for (var re : r.entrySet())
                    nxt.merge(re.getKey(), pe.getValue() * re.getValue(), Long::sum);
            }
            cur = nxt;
            if (cur.isEmpty()) break;
        }
        return cur;
    }
}

File: everything_parsing/29-llk-adaptive/js/backtrack.mjs
// Module 29 — a backtracking recursive-descent recognizer that COUNTS
// (mirror of java/Backtrack.java).
//
// deriveSym tries every production, follows all of them (no lookahead), and
// returns per-end-position the number of ways to derive the consumed span.
// `entries` = production attempts = the raw work. No memoization -> exponential
// when alternatives overlap. Packrat (Appendix X2) memoizes (nt, pos) -> linear.
// Grammar must be free of left recursion.

export class Backtrack {
  constructor(g, cap = 20_000_000) { this.g = g; this.cap = cap; this.entries = 0; }

  parse(input) {
    this.entries = 0;
    try {
      const ends = this.deriveSym(this.g.start, 0, input);
      return { parseCount: ends.get(input.length) ?? 0, entries: this.entries, capped: false };
    } catch (e) {
      if (e === "capped") return { parseCount: -1, entries: this.entries, capped: true };
      throw e;
    }
  }

  deriveSym(sym, pos, input) {
    if (!this.g.isNT(sym)) {
      const out = new Map();
      if (pos < input.length && input[pos] === sym) out.set(pos + 1, 1);
      return out;
    }
    const out = new Map();
    for (const rhs of this.g.prods.get(sym)) {
      if (++this.entries > this.cap) throw "capped";
      const r = this.deriveSeq(rhs, pos, input);
      for (const [e, w] of r) out.set(e, (out.get(e) ?? 0) + w);
    }
    return out;
  }

  deriveSeq(seq, pos, input) {
    let cur = new Map([[pos, 1]]);
    for (const sym of seq) {
      const nxt = new Map();
      for (const [p, ways] of cur) {
        const r = this.deriveSym(sym, p, input);
        for (const [e, w] of r) nxt.set(e, (nxt.get(e) ?? 0) + ways * w);
      }
      cur = nxt;
      if (cur.size === 0) break;
    }
    return cur;
  }
}
